### DISCUSSION TITLE
How do I create a single-point run with the NUOPC (default) coupler?

### DISCUSSION THREAD

**Discussion by adrifoster** [ESCOMP/CTSM#1833] [Q&A] [answered]
We get this question a lot:

I am trying to set up a case for a specific latitude and longitude. I want to use GSWP3 climate data, how do I set this up?

**Comment by adrifoster** [comment, marked as answer]
We now have a method for subsetting domain, surface, and DATM data (and soon meshes - for regional cases) which you can use to set up single-point cases.

We recently went over how to do this at the 2022 CTSM Tutorial - that notebook can be found [here](https://github.com/NCAR/CTSM-Tutorial/blob/Mini-Tutorial-2022/notebooks/Day2a_GenericSinglePoint.ipynb). But see below for some general instructions.

------------------------------
## General Instructions 

### Subsetting the data
For single-point cases with NUOPC, you need to subset a surface dataset and (optionally) DATM data.  The python script to subset this data can be found in the CTSM repository at [**tools/site_and_regional/subset_data**](https://github.com/adrifoster/CTSM/blob/master/tools/site_and_regional/subset_data). 

Note that you will need to have a python environment set up that includes the packages **scipy**, **xarray,** and **numpy**. If you have conda or miniconda installed, you can create a conda environment for this and other CTSM python tools using the script [`./py_env_create` ](https://github.com/adrifoster/CTSM/blob/master/py_env_create).

To subset surface and DATM data for a single point, use the command:

```
./subset_data point --lat $my_lat --lon $my_lon --site $my_site_name --create-surface --create-datm --datm-syr $my_start_year --datm-eyr $my_end_year --create-user-mods --outdir $my_output_dir
```

`$my_lat`              : latitude of point, *should be between -90 and 90 degrees*
`$my_lon`             : longitude of point, *should be between 0 and 360 or -180 and 180 degrees*
`$my_site_name` : name of site, *used for file naming*
`$my_start_year` : start year for DATM data to subset, *default between 1901 and 2014*
`$my_end_year`   : end year for DATM data to subset, *default between 1901 and 2014*
`$my_output_dir` : output directory to place the subset data and user_mods directory

You can also have the script subset landuse data. See the help (`./subset_data --help`) for all argument options.

**_Also note that this script defaults to subsetting specific (default) surface, domain, and landuse files and GSWP3 DATM data, and can currently only be run as-is on Cheyenne. To update the files and file locations, you will need to modify the files and/or directories in the [`deafult_data.cfg`](https://github.com/adrifoster/CTSM/blob/master/tools/site_and_regional/default_data.cfg) file._**

The `--create-user-mods` command tells the script to set up a user mods directory in your specified `$my_output_dir` and to specify the required `PTS_LAT` and `PTS_LON` parameters. You can then use this user mods directory to kick off your CTSM case.

### Creating the case

With the default NUOPC compiler, you can use the user mods directory set up in the previous subset data step to tell CIME/CTSM where your subset files are located. 

Navigate to the CTSM _cime/scripts_ directory to create the case:

```
cd ~/CTSM/cime/scripts
./create_newcase --case $my_case_name --res CLM_USRDAT --compset $compset --run-unsupported --user-mods-dirs $my_output_dir/user_mods
```

`$my_case_name` : the name/location of your case
`$compset`            : the compset you would like to use (see the [above-mentioned tutorial](https://github.com/NCAR/CTSM-Tutorial-2022/blob/main/notebooks/Day2a_GenericSinglePoint.ipynb) for an example)
Note the use of `$my_output_dir/user_mods` which is the **user_mods** directory that the subset data script set up within your specified `$my_output_dir`.

Following this, you should be able to update any other case-specific parameters you want to change (e.g., `STOP_N`, `STOP_OPTION`, etc.). You should set the values `DATM_YR_ALIGN`, `DATM_YR_START`, and `DATM_YR_END` to match your `$my_end_year` and `$my_start_year` values from above:

```
./xmlchange DATM_YR_ALIGN=$my_start_year
./xmlchange DATM_YR_START=$my_start_year
./xmlchange DATM_YR_END=$my_end_year
```

**Reply by daofeng2007** [reply]
Hi Adrianna, thank you for your tutorial. I tried general point simulation on my own computer and I got stuck in the "./xmlchange MPILIB=impi" step. If I implemented "./xmlchange MPILIB=impi",  I received the error message "ERROR: Could not find a matching MPI for attributes: {'compiler': 'gnu', 'mpilib': 'impi', 'threaded': False}". If I skip that "./xmlchange MPILIB=impi" line, the case.build was successful, but I received the error below in the case.submit step. I am using escomp/cesm-lab-neon, and escomp/ctsm-neon containers to do the experiment and I can run global simulation in those containers.

*** The MPI_Query_thread() function was called before MPI_INIT was invoked.
*** This is disallowed by the MPI standard.
*** Your MPI job will now abort.
[cc1348ebea6c:34309] Local abort before MPI_INIT completed completed successfully, but am not able to aggregate error messages, and not able to guarantee that all other processes were killed!

**Comment by olyson** [comment]
I tried this using ctsm5.1.dev118:

./subset_data point --lat 40.0 --lon 270.0 --site SITE_TEST --create-surface --create-datm --datm-syr 2000 --datm-eyr 2000 --create-user-mods --outdir /glade/scratch/oleson/SITE_TEST

It completed without errors.  However, the time dimension in the Solar file appears to be wrong (truncated), e.g.,

ncdump -v time /glade/scratch/oleson/SITE_TEST/datmdata/clmforc.GSWP3.c2011.0.5x0.5.Solr.SITE_TEST.2000-01.nc

 time = 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2,
    2, 3, 3, 3, 3, 3, 3, 3, 3, 4, 4, 4, 4, 4, 4, 4, 4, 5, 5, 5, 5, 5, 5, 5,
    5, 6, 6, 6, 6, 6, 6, 6, 6, 7, 7, 7, 7, 7, 7, 7, 7, 8, 8, 8, 8, 8, 8, 8,
    8, 9, 9, 9, 9, 9, 9, 9, 9, 10, 10, 10, 10, 10, 10, 10, 10, 11, 11, 11,
    11, 11, 11, 11, 11, 12, 12, 12, 12, 12, 12, 12, 12, 13, 13, 13, 13, 13,
    13, 13, 13, 14, 14, 14, 14, 14, 14, 14, 14, 15, 15, 15, 15, 15, 15, 15,
    15, 16, 16, 16, 16, 16, 16, 16, 16, 17, 17, 17, 17, 17, 17, 17, 17, 18,
    18, 18, 18, 18, 18, 18, 18, 19, 19, 19, 19, 19, 19, 19, 19, 20, 20, 20,
    20, 20, 20, 20, 20, 21, 21, 21, 21, 21, 21, 21, 21, 22, 22, 22, 22, 22,
    22, 22, 22, 23, 23, 23, 23, 23, 23, 23, 23, 24, 24, 24, 24, 24, 24, 24,
    24, 25, 25, 25, 25, 25, 25, 25, 25, 26, 26, 26, 26, 26, 26, 26, 26, 27,
    27, 27, 27, 27, 27, 27, 27, 28, 28, 28, 28, 28, 28, 28, 28, 29, 29, 29,
    29, 29, 29, 29, 29, 30, 30, 30, 30, 30, 30, 30, 30 ;

Is this a known problem?

**Reply by ekluzek** [reply]
@olyson we should move this to its own issue. But, isn't 248 time-samples for January the correct number?

**Reply by olyson** [reply]
Sorry I shouldn't have said time dimension.  The time values themselves are truncated, e.g., the first few values are 0,0,0,0,0, etc.

**Reply by ekluzek** [reply]
Ahh, it's converting to integer somehow, and only for Solar and not for Prec or TPQWL. The coordinate variable itself is float, so it's not obvious why this would happen for Solar and not the other two.

**Comment by wwieder** [comment]
Working with ctsm5.3.012 now and needed to set `./xmlchange DATM_MODE=CLMGSWP3v1` for the user_nl_datm_streams to work correctly now.  did something change in cdeps or our namelist settings that changed this?

**Reply by samsrabin** [reply]
I don't know, but this makes me think we should have a system test for this...

**Reply by wwieder** [reply]
then I'd be out of a job of finding broken bits in our single point workflows?!  Good idea @samsrabin

**Comment by samsrabin** [comment]
The most up-to-date instructions (based in part on your answer here, thanks!) are now [here](https://escomp.github.io/ctsm-docs/versions/master/html/users_guide/running-single-points/running-single-point-subset-data.html). @adrifoster, could you mark this comment as best answer, then close this discussion?
